<a href="https://colab.research.google.com/github/sebekpro/python-krakowiak/blob/main/250-data-science-bootcamp-w-jezyku-python/211_220_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Data Science Bootcamp

### Spis treści:
* [Import biblioteki](#0)
* [Ćwiczenie 211](#1)
* [Ćwiczenie 212](#2)
* [Ćwiczenie 213](#3)
* [Ćwiczenie 214](#4)
* [Ćwiczenie 215](#5)
* [Ćwiczenie 216](#6)
* [Ćwiczenie 217](#7)
* [Ćwiczenie 218](#8)
* [Ćwiczenie 219](#9)
* [Ćwiczenie 220](#10)

### <a name='0'></a>Import biblioteki

In [ ]:
import numpy as np
import pandas as pd

np.__version__

'1.26.4'

### <a name='1'></a> Ćwiczenie 211
Podany jest poniższy obiekt _DataFrame_:

```
df = pd.DataFrame(data={'height': [175., 178.5, 185., 191., 184.5, 183., 168.]})
```
Dokonaj dyskretyzacji zmiennej _height_ na 3 przedziały o równej szerokości. Wynik przypisz do nowej kolumny _height_cut_.


In [ ]:
# tutaj wpisz rozwiązanie
df = pd.DataFrame(data={'height': [175., 178.5, 185., 191., 184.5, 183., 168.]})

# Dyskretyzacja zmiennej 'height' na 3 przedziały o równej szerokości
df['height_cut'] = pd.cut(df['height'], bins = 3)

df

height          height_cut
0   175.0  (167.977, 175.667]
1   178.5  (175.667, 183.333]
2   185.0    (183.333, 191.0]
3   191.0    (183.333, 191.0]
4   184.5    (183.333, 191.0]
5   183.0  (175.667, 183.333]
6   168.0  (167.977, 175.667]

Funkcja pd.cut dzieli ciągły zakres wartości zmiennej na określoną liczbę przedziałów o równej szerokości. W przypadku zmiennej height oraz argumentu bins=3:

1. Obliczanie zakresu wartości:
Funkcja wyznacza minimalną i maksymalną wartość w kolumnie height. Dla przykładu, jeśli mamy wartości od 168.0 do 191.0, to zakres wynosi 23.0.

2. Wyznaczenie szerokości przedziału:
Całkowity zakres jest dzielony przez liczbę przedziałów. W naszym przykładzie szerokość jednego przedziału wynosi 23.0/0 ≈ 7.67

3. Ustalenie granic przedziałów:
Na podstawie minimalnej wartości oraz szerokości przedziału funkcja automatycznie określa granice:
* Pierwszy przedział: od 168.0 do 168.0 + 7.67 ≈ 175.67,
* Drugi przedział: od 175.67 do 175.67 + 7.67 ≈ 183.33,
* Trzeci przedział: od 183.33 do 191.0.

Domyślnie przedziały są tworzone jako otwarte z lewej i domknięte z prawej stroną, czyli np. (168.0, 175.67].

4. Przypisanie wartości do przedziałów:
Każda wartość w kolumnie height jest przyporządkowywana do odpowiedniego przedziału, w którym się znajduje. W wyniku tej operacji otrzymujemy zmienną kategoryczną, gdzie każda obserwacja ma etykietę przedziału (np. (175.67, 183.33]).

Podsumowując, dyskretyzacja zmiennej height przy użyciu pd.cut(height, bins=3) polega na automatycznym podziale zakresu wartości na 3 równe segmenty, a następnie przypisaniu każdej wartości do odpowiadającego jej segmentu. Wynik jest nową kolumną, która kategoryzuje wartości zmiennej height według ustalonych przedziałów.

### <a name='2'></a> Ćwiczenie 212
Dokonaj dyskretyzacji zmiennej _height_ na 3 przedziały o zadanej postaci:
* (160, 175]
* (175, 180]
* (180, 195]

Wynik przypisz do nowej kolumny _height_cut_2_.

In [ ]:
# tutaj wpisz rozwiązanie

# Definicja przedziałów
bins = [160, 175, 180, 195]

# Dyskretyzacja zmiennej 'height' przy użyciu zdefiniowanych przedziałów
df['height_cut_2'] = pd.cut(df['height'], bins = bins)

df

height          height_cut height_cut_2
0   175.0  (167.977, 175.667]   (160, 175]
1   178.5  (175.667, 183.333]   (175, 180]
2   185.0    (183.333, 191.0]   (180, 195]
3   191.0    (183.333, 191.0]   (180, 195]
4   184.5    (183.333, 191.0]   (180, 195]
5   183.0  (175.667, 183.333]   (180, 195]
6   168.0  (167.977, 175.667]   (160, 175]

### <a name='3'></a> Ćwiczenie 213
Dokonaj dyskretyzacji zmiennej _height_ na 3 przedziały o zadanej postaci:
* (160, 175]
* (175, 180]
* (180, 195]

oraz przypisz im odpowiednio etykiety:

* small
* medium
* high

Wynik przypisz do nowej kolumny _height_cut_3_.

In [ ]:
# tutaj wpisz rozwiązanie

### <a name='4'></a> Ćwiczenie 214
Usuń kolumny _height_cut_ oraz _height_cut_2_ z obiektu _df_.

In [ ]:
# tutaj wpisz rozwiązanie

Zmień nazwę kolumny _height_cut_3_ na _height_level_.

In [ ]:
# tutaj wpisz rozwiązanie

### <a name='5'></a> Ćwiczenie 215
Uzyskaj kodowanie 0-1 obiektu _df_ dzięki funkcji _pd.get_dummies()_.

In [ ]:
# tutaj wpisz rozwiązanie

### <a name='6'></a> Ćwiczenie 216
Podany jest poniższy obiekt _df_:


```
df = pd.DataFrame(data={'lang': [['PL', 'ENG'], ['GER', 'ENG', 'PL', 'FRA'], ['RUS']]})
```
Zbuduj nową kolumnę o nazwie _length_, która przyjmie długość listy kolumny _lang_.



In [ ]:
# tutaj wpisz rozwiązanie

### <a name='7'></a> Ćwiczenie 217
Zbuduj nową kolumnę o nazwie _PL_flag_, która przyjmie wartość 1, gdy język 'PL' będzie w liście w kolumnie _lang_, przeciwnie 0.

In [ ]:
# tutaj wpisz rozwiązanie

### <a name='8'></a> Ćwiczenie 218
Podany jest poniższy obiekt _df_:


```
df = pd.DataFrame(data={'website': ['wp.pl', 'onet.pl', 'google.com']})
```
Podziel kolumnę _website_ względem znaku kropki (.) na dwie kolumny o nazwach _portal_ oraz _ending_.


In [ ]:
# tutaj wpisz rozwiązanie

### <a name='9'></a> Ćwiczenie 219
Wytnij wszystkie wiersze obiektu _df_ dla którego kolumna _portal_ zawiera literę 'o'.

In [ ]:
# tutaj wpisz rozwiązanie

### <a name='10'></a> Ćwiczenie 210
Zamień wszystkie rozszerzenia _com_ z kolumny _ending_ na _org_.

In [ ]:
# tutaj wpisz rozwiązanie